[MLIR][Linalg] Pattern to fold AddOp to accumulation via contraction op's dest

rolfmorel opened the pull request:
Replaces a linalg.add with one operand the single user of a contraction, which has a zero-filled, "identity-mapped" destination and is dominated by the `other` operand, by the contraction with `other` as its dest.

Benefits include elision of an elementwise op, namely the linalg.add, and removing a tensor.empty as a destination which is likely to require an allocation upon bufferization.

Commit:
[MLIR][Linalg] Pattern to fold AddOp to accumulation via contraction op's dest

Replaces a linalg.add with one operand the single user of a contraction,
which has a zero-filled, "identity-mapped" destination and is dominated by
the `other` operand, by the contraction with `other` as its dest.

Benefits include elision of an elementwise op, namely the linalg.add,
and removing a tensor.empty as a destination which is likely to require an
allocation upon bufferization.

rolfmorel commented:
Credit for the `isZero` utility functions goes to @rengolin and @chelini.

rolfmorel commented:
Original downstream PR is here: https://github.com/plaidml/tpp-mlir/pull/966

Credit to @adam-smnk for suggesting the transform and thanks for the thorough review on the original PR!

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/FoldAddIntoDest.cpp:
```diff
@@ -0,0 +1,107 @@
+      //     region there is no need to perform checks on addOp's out argument.
+    }
+
+    // Dominated op must be a contraction for it to accumulate on its out arg.
+    // E.g., AddOp is not a contraction and hence ignores its out arg's value.
```
nit: there's probably whole class of generics that would fit this folding but it's perfectly fine to restrict this to contraction, perhaps I'd just slightly generalize this comment

rolfmorel replied:
True - have added a TODO comment regarding this.

banach-space commented on mlir/test/Dialect/Linalg/fold-add-into-dest.mlir:
```diff
@@ -0,0 +1,288 @@
+// RUN: mlir-opt %s -transform-interpreter -cse -split-input-file | FileCheck %s
+
+!type = tensor<2048x2048xf32>
+func.func @expect_add_to_fold(%arg0: !type, %arg1: !type) -> !type {
```
[nit] Could we have more descriptive function names? In particular, what's the difference between this version of `@expect_add_to_fold` and the one below?

rolfmorel replied:
Fair - have changed the names to reflect the differences in test cases.

banach-space commented on mlir/test/Dialect/Linalg/fold-add-into-dest.mlir:
```diff
@@ -0,0 +1,288 @@
+
+// CHECK-LABEL: func.func @expect_add_to_fold
+// CHECK: %[[ACC:.+]] = linalg.matmul
+// CHECK-NEXT: %[[RES:.+]] = linalg.matmul ins({{.+}}) outs(%[[ACC]]
```
CHECK-NOT: linalg.add?

rolfmorel replied:
Have added it (though as in every instance it was a `CHECK-NEXT: return` on the next line, I believe it shouldn't make a difference).

banach-space replied:
Sorry, I missed `CHECK-NEXT`. You are right, `CHECK-NOT` is not really needed in this case.

banach-space commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -870,5 +870,80 @@ getReassociationMapForFoldingUnitDims(ArrayRef<OpFoldResult> mixedSizes) {
+      .Case<DenseElementsAttr>([](auto attr) {
+        if (!attr.getElementType().isIntOrFloat())
+          return false;
+        if (!attr.isSplat())
+          return false;
+        auto splat = attr.template getSplatValue<Attribute>();
+        return isZeroAttr(splat);
```
Why do you need to recurse into `isZeroAttr` in this case? Also:
```suggestion
        // Splat of 0
        if (attr.isSplat() && (attr.getSplatValue<APInt>().isZero() || attr.getSplatValue<APFloat>().isZero())
          return true;
```

Hope I didn't miss anything myself.

rolfmorel replied:
Turns out the "predicate matcher" function `m_AnyZeroFloat` and `m_Zero` already deal with splatted zeroes.

Have made all this code substantially simpler.

banach-space commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -870,5 +870,80 @@ getReassociationMapForFoldingUnitDims(ArrayRef<OpFoldResult> mixedSizes) {
   return reassociation;
 }
 
+// Returns true if the value is a constant float or integer.
```
Constant or "zero"? Also, is this only meant for scalars?

rolfmorel replied:
Removed this.

banach-space commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -870,5 +870,80 @@ getReassociationMapForFoldingUnitDims(ArrayRef<OpFoldResult> mixedSizes) {
+
+// Recurses into isZeroOp for defining ops if not immediately obvious.
+// Looks past linalg generic's argument (which don't have defining ops).
+bool isZeroTensor(Value val) {
```
This method is called `isZeroTensor`, but it recurses to `isZeroOp`, which also works for Memref Ops. I don't see any tests with Memrefs and I feel that all of this can safely bail out in the case of Memrefs. WDTY?

banach-space commented on mlir/include/mlir/Dialect/Linalg/Utils/Utils.h:
```diff
@@ -38,6 +38,12 @@ namespace linalg {
+bool isZeroTensor(Value val);
+
+// Returns true if the operation defines a zero-filled tensor.
+bool isZeroOp(Operation *);
```
[nit] I'd try a bit more descriptive name
```suggestion
bool isZeroLikeOp(Operation *);
```

Also, this is not really used at the Linalg level, is it? It's only needed in Utils.cpp, so I'd skip declaring it here.

rolfmorel replied:
I've removed this.

banach-space commented on mlir/include/mlir/Dialect/Linalg/Utils/Utils.h:
```diff
@@ -38,6 +38,12 @@ namespace linalg {
 //===----------------------------------------------------------------------===//
 
+// Returns true if `val` represents a zero-filled tensor, per its defining op.
+bool isZeroTensor(Value val);
```
This is a fairly generic hook and nothing specific to Linalg, is it?

rolfmorel replied:
As it has support for tracing a value's definition through `linalg.generic`s, `linalg.fill` and `linalg.copy` it is kind of tied to linalg, I would say. As such I didn't think the code could be moved to StaticValueUtils.cpp (which is pretty much dialect/op agnostic).

banach-space commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -870,5 +870,80 @@ getReassociationMapForFoldingUnitDims(ArrayRef<OpFoldResult> mixedSizes) {
+      .Case<memref::CopyOp, memref::SubViewOp, tensor::CastOp,
+            tensor::ExtractSliceOp>(
+          [&](auto op) { return isZeroTensor(op.getSource()); })
+      .Case<memref::GetGlobalOp>([&](auto op) {
```
Is there a test for this case?

banach-space commented on mlir/lib/Dialect/Linalg/Utils/Utils.cpp:
```diff
@@ -870,5 +870,80 @@ getReassociationMapForFoldingUnitDims(ArrayRef<OpFoldResult> mixedSizes) {
+          return false;
+        return isZeroTensor(op.getInputs()[0]);
+      })
+      .Case<memref::CopyOp, memref::SubViewOp, tensor::CastOp,
```
Are there any tests for the memref Ops?

rolfmorel replied:
Indeed, this code was not tested. Have now clearly delineated this PR to be about tensors only and removed the above code.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/FoldAddIntoDest.cpp:
```diff
@@ -0,0 +1,109 @@
+
+using namespace mlir;
+
+/// Replace a linalg.add with one operand the single user of a contraction,
+/// which has a zero-filled, "identity-mapped" destination and is dominated by
+/// the `other` operand, by the contraction with `other` as its dest.
```
Could you add a simple (potentially pseudo) example as well? As in, "BEFORE" and "AFTER" example.

banach-space reviewed (commented):
The helpers that you added in Linalg/Utils/Utils.h are not really specific to Linalg and should be moved to somewhere more general. Given the [location](StaticValueUtils.h) of `isZeroIndex`, I suggest StaticValueUtils.h. Also, it feels like some of the logic that you added there should already exist 🤔 

I also suggest bailing out early when encountering memrefs. Unless that's something that this is meant to support?

banach-space commented on mlir/test/Dialect/Linalg/fold-add-into-dest.mlir:
```diff
@@ -0,0 +1,291 @@
+}
+
+// CHECK-LABEL: func.func @fold_add_on_two_matmuls
+// CHECK: %[[ACC:.+]] = linalg.matmul
```
It would be good to have at least a few tests where also types (and possibly other details) are verified.

rolfmorel replied:
Have converted one test case to now include full details regarding arguments. I have left out types as only arguments get shuffled around and their shapes ought to be same before the rewrite already.

banach-space commented on mlir/test/Dialect/Linalg/fold-add-into-dest.mlir:
```diff
@@ -0,0 +1,291 @@
+// -----
+
+!type = tensor<2048x2048xf32>
+func.func @expect_no_fold_as_orig_dest_not_additive_zero(%arg0: !type, %arg1: !type) -> !type {
```
This is very similar to `@fold_add_on_two_matmuls`. I think that it would be beneficial to group all similar tests together so that it's clear what edge cases have been exercised. To me, this would make sense:
* `@expect_no_fold_as_orig_dest_not_additive_zero` -> `@expect_no_fold_add_on_matmuls_orig_dest_not_additive_zero`

and then move near `@fold_add_on_two_matmuls`. Similar comment for other tests.

This is a non-blocker, just a kind request :) It makes maintaining tests much easier - I'm learning the hard way while going through Vector dialect tests 😅 Having said all that, these tests are already very nicely formatted, thanks!

banach-space replied:
> Note that previously I had them grouped by "interesting cases that ought to be working" and a group of "a test for each early exit condition". I guess its matter of personal preference which grouping is more helpful.

Good point, I'm clearly biased towards my favourite my approach :) Your approach also works, but then adding a big  comment separating "positive" from "negative" tests would help. It wasn't obvious to me that that's how you organised them.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/FoldAddIntoDest.cpp:
```diff
@@ -0,0 +1,152 @@
+    Value dominatingOperand = nullptr;
+    linalg::LinalgOp dominatedOp = nullptr;
+    {
+      auto firstOperand = addOp.getOperand(0);
+      auto secondOperand = addOp.getOperand(1);
```
[nit] I always advocate for long descriptive names, but in this case short `rhs` and `lhs` might be even more descriptive :)

Could you spell out `auto`?

rolfmorel replied:
Agreed - and done.

banach-space commented on mlir/lib/Dialect/Linalg/Transforms/FoldAddIntoDest.cpp:
```diff
@@ -0,0 +1,152 @@
+                                PatternRewriter &rewriter) const override {
+    Value dominatingOperand = nullptr;
+    linalg::LinalgOp dominatedOp = nullptr;
+    {
```
Why the extra braces?

rolfmorel replied:
I wanted to limit the scope of the variables `lhs` and `rhs` as after this block all that matters are the values of `dominatingOperand` and `dominatedOp`. It is a way to keep myself/the code after this block to the contract that the order of the operands shouldn't matter.

Have added a comment on the block to explain.

banach-space commented:
> If you could have another look at the now simpler code, that would be much appreciated.

I really like how you simplified that, thanks!

> As for where utility functions like this should exist, I am interested as well if the functionality already exists somewhere. I have not been able to find it so far (which ties into the question what the right location would be for such functions). 😢

I couldn't anything myself. Lets just be mindful of this and make sure we don't re-implement this anywhere - that wouldn't be the first time 😂

banach-space reviewed (approved):
LGTM, thanks for addressing my comments! My remaining suggestions are nits/nice-to-haves.

Please wait for one more +1 before landing. @adam-smnk , perhaps you could take another look?

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/FoldAddIntoDest.cpp:
```diff
@@ -0,0 +1,152 @@
+      auto secondOperand = addOp.getOperand(1);
+
+      // For now, pattern only applies to tensor types (memref support is TODO).
+      if (!isa<TensorType>(addOp->getResult(0).getType()) ||
```
This will crash on memrefs, maybe you could call `hasPureTensorSemantics` on the addOp?

adam-smnk replied:
And could you add at least one negative test case on memrefs?

rolfmorel replied:
Good catch. Have switched to `hasPureTensorSemantics` and added a negative memref test case.

adam-smnk commented on mlir/lib/Dialect/Linalg/Transforms/FoldAddIntoDest.cpp:
```diff
@@ -0,0 +1,152 @@
+    int prevDimPos = -1;
+    for (auto expr : indexMaps[destOperand->getOperandNumber()].getResults()) {
+      auto dim = dyn_cast<AffineDimExpr>(expr);
+      if (!dim || prevDimPos >= (int)dim.getPosition())
```
nit: static_cast

rolfmorel replied:
Done.

adam-smnk reviewed (approved):
Looks good, thanks 👍

rolfmorel commented:
Thank you for the comments, @adam-smnk and @banach-space - it really helped clean-up the PR!

I just now rebased, squashed and ran a `check-all`. The PR is ready to merge, I say. If either of you could help with merging, that would be great!
